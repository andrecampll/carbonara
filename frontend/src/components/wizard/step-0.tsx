import { Button, Typography } from '@mui/material'
import Image from 'next/image'
import { useWizard } from 'react-use-wizard'

import { Step } from './step'

export function Step0() {
  const { previousStep, nextStep } = useWizard()

  return (
    <Step>
      <div className="flex flex-col gap-4">
        <header>
          <Typography variant="h5">
            Lets Calculate Your Carbon Footprint
          </Typography>
          <Typography fontWeight={400}>
            Follow the steps to calculate your carbon footprint. We will provide
            you with a detailed report at the end. Lets get started!
          </Typography>
        </header>

        <div className="my-4 flex flex-col items-center gap-4">
          <Image
            alt=""
            width={200}
            height={200}
            src="/images/carbon-light.webp"
          />
        </div>

        <footer className="flex w-full items-center justify-between">
          <Button onClick={previousStep} color="secondary" type="submit">
            Cancel
          </Button>
          <Button
            onClick={nextStep}
            variant="contained"
            color="secondary"
            type="submit"
          >
            Next
          </Button>
        </footer>
      </div>
    </Step>
  )
}
