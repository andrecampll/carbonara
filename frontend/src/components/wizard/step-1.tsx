import { zodResolver } from '@hookform/resolvers/zod'
import { Button, Typography } from '@mui/material'
import Image from 'next/image'
import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { useWizard } from 'react-use-wizard'
import { z } from 'zod'

import { Input } from '../ui/input'
import { useFootprintWizard } from './hooks/use-footprint-wizard'

const footprintStep1Schema = z.object({
  electrictyUsage: z.string().min(1, {
    message: 'Electricity usage is required',
  }),
})

type FootprintWizardFormStep1Data = z.infer<typeof footprintStep1Schema>

export function Step1() {
  const { previousStep, nextStep } = useWizard()
  const { updateWizard, wizard } = useFootprintWizard()

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FootprintWizardFormStep1Data>({
    resolver: zodResolver(footprintStep1Schema),
    defaultValues: {
      electrictyUsage: wizard?.electrictyUsage?.toString() ?? '',
    },
  })

  const onSubmit = useCallback(
    async (formData: FootprintWizardFormStep1Data) => {
      // handle submit
      updateWizard({
        electrictyUsage: Number(formData.electrictyUsage),
      })
      nextStep()
    },
    [nextStep, updateWizard],
  )

  return (
    <div className="flex flex-col gap-4">
      <header>
        <Typography variant="h5">Electricity Usage</Typography>
        <Typography fontWeight={400}>
          Enter your monthly electricity usage in kilowatt hours (Kwh) to
          calculate your carbon footprint.
        </Typography>
      </header>

      <form
        onSubmit={handleSubmit(onSubmit)}
        className="my-4 flex flex-col items-center gap-4"
      >
        <Image alt="" width={300} height={300} src="/images/electricity.webp" />

        <Input
          label="Electricity Usage (Kwh/Month):"
          type="number"
          helperText={errors.electrictyUsage?.message}
          color="secondary"
          {...register('electrictyUsage')}
        />

        <footer className="flex w-full items-center justify-between">
          <Button onClick={previousStep} color="secondary">
            Back
          </Button>
          <Button variant="contained" color="secondary" type="submit">
            Next
          </Button>
        </footer>
      </form>
    </div>
  )
}
