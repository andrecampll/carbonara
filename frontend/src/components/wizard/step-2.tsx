import { zodResolver } from '@hookform/resolvers/zod'
import { Button, Typography } from '@mui/material'
import Image from 'next/image'
import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { useWizard } from 'react-use-wizard'
import { z } from 'zod'

import { Input } from '../ui/input'
import { useFootprintWizard } from './hooks/use-footprint-wizard'
import { Step } from './step'

const footprintStep2Schema = z.object({
  transportationUsage: z.string().min(1, {
    message: 'Transportation gasoline usage is required',
  }),
})

type FootprintWizardFormStep2Data = z.infer<typeof footprintStep2Schema>

export function Step2() {
  const { previousStep, nextStep } = useWizard()
  const { updateWizard, wizard } = useFootprintWizard()

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FootprintWizardFormStep2Data>({
    resolver: zodResolver(footprintStep2Schema),
    defaultValues: {
      transportationUsage: wizard?.transportationUsage?.toString() ?? '',
    },
  })

  const onSubmit = useCallback(
    async (formData: FootprintWizardFormStep2Data) => {
      updateWizard({
        transportationUsage: Number(formData.transportationUsage),
      })
      nextStep()
    },
    [nextStep, updateWizard],
  )

  return (
    <Step>
      <div className="flex flex-col gap-4">
        <header>
          <Typography variant="h5">Transportation Gasoline Usage</Typography>
          <Typography fontWeight={400}>
            Enter your monthly transportation gasoline usage in gallons to
            calculate your carbon footprint.
          </Typography>
        </header>

        <form
          onSubmit={handleSubmit(onSubmit)}
          className="my-4 flex flex-col items-center gap-4"
        >
          <Image alt="" width={300} height={300} src="/images/gas.png" />

          <Input
            label="Transportation Gasoline Usage (Gallons/Month):"
            type="number"
            helperText={errors.transportationUsage?.message}
            {...register('transportationUsage')}
          />

          <footer className="flex w-full items-center justify-between">
            <Button onClick={previousStep} color="secondary">
              Back
            </Button>
            <Button variant="contained" color="secondary" type="submit">
              Next
            </Button>
          </footer>
        </form>
      </div>
    </Step>
  )
}
