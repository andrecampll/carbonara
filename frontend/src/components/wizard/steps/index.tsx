'use client'

import { AnimatePresence } from 'framer-motion'
import { ReactNode } from 'react'
import { Wizard } from 'react-use-wizard'

type Props = {
  children: ReactNode
}

export function Steps({ children }: Props) {
  return (
    <Wizard wrapper={<AnimatePresence initial={false} mode="wait" />}>
      {children}
    </Wizard>
  )
}
