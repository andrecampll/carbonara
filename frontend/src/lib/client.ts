import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client'
import { registerApolloClient } from '@apollo/experimental-nextjs-app-support/rsc'

import { env } from '@/env'

export const { getClient } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: env.NEXT_PUBLIC_API_BASE_URL,
    }),
  })
})
