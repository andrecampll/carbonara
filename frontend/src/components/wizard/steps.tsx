'use client'

import { useMutation } from '@apollo/client'
import { useCallback } from 'react'
import { Wizard } from 'react-use-wizard'

import {
  CalculateFootprintMutationInput,
  Footprint,
} from '@/data/types/footprint'
import { CALCULATE_FOOTPRINT_MUTATION } from '@/graphql/mutations/calculateFootprint'
import { useFootprint } from '@/hooks/use-footprint'

import { Step0 } from './step-0'
import { Step1 } from './step-1'
import { Step2 } from './step-2'
import { Step3 } from './step-3'
import { Step4 } from './step-4'

type CalculateFootprintMutationResponse = {
  calculate: {
    footprint: Footprint
  }
}

export function Steps() {
  const { addFootprint } = useFootprint()

  const [calculateFunction] = useMutation<CalculateFootprintMutationResponse>(
    CALCULATE_FOOTPRINT_MUTATION,
  )

  const handleCalculateFootprint = useCallback(
    async (wizardData: CalculateFootprintMutationInput['input']) => {
      const { data: footPrintData } = await calculateFunction({
        variables: {
          input: wizardData,
        },
      })

      addFootprint(footPrintData?.calculate.footprint as Footprint)
    },
    [calculateFunction, addFootprint],
  )

  return (
    <Wizard>
      <Step0 />
      <Step1 />
      <Step2 />
      <Step3 />
      <Step4 onFinalSubmit={handleCalculateFootprint} />
    </Wizard>
  )
}
