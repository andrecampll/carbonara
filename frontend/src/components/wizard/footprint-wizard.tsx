import { useMutation } from '@apollo/client'
import { Card, CardContent } from '@mui/material'
import { useCallback } from 'react'

import {
  CalculateFootprintMutationInput,
  Footprint,
} from '@/data/types/footprint'
import { CALCULATE_FOOTPRINT_MUTATION } from '@/graphql/mutations/calculateFootprint'
import { useFootprint } from '@/hooks/use-footprint'
import { FootprintWizardProvider } from '@/hooks/use-footprint-wizard'

import { Steps } from './steps'
import { Step0 } from './steps/step-0'
import { Step1 } from './steps/step-1'
import { Step2 } from './steps/step-2'
import { Step3 } from './steps/step-3'
import { Step4 } from './steps/step-4'

type CalculateFootprintMutationResponse = {
  calculate: {
    footprint: Footprint
  }
}

type Props = {
  startIndex?: number
}

export function FootprintWizard({ startIndex }: Props) {
  const { addFootprint } = useFootprint()

  const [calculateFunction] = useMutation<CalculateFootprintMutationResponse>(
    CALCULATE_FOOTPRINT_MUTATION,
  )

  const handleCalculateFootprint = useCallback(
    async (wizardData: CalculateFootprintMutationInput['input']) => {
      const { data: footPrintData } = await calculateFunction({
        variables: {
          input: wizardData,
        },
      })

      addFootprint(footPrintData?.calculate.footprint as Footprint)
    },
    [calculateFunction, addFootprint],
  )

  return (
    <div className="w-full">
      <Card>
        <CardContent>
          <FootprintWizardProvider>
            <Steps startIndex={startIndex}>
              <Step0 />
              <Step1 />
              <Step2 />
              <Step3 />
              <Step4 onFinalSubmit={handleCalculateFootprint} />
            </Steps>
          </FootprintWizardProvider>
        </CardContent>
      </Card>
    </div>
  )
}
