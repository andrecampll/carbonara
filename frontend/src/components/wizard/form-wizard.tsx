import { Card, CardContent } from '@mui/material'

import { WizardProvider } from '@/hooks/use-footprint-wizard'

import { Steps } from './steps'

export function FormWizard() {
  return (
    <div className="w-full">
      <Card>
        <CardContent>
          <WizardProvider>
            <Steps />
          </WizardProvider>
        </CardContent>
      </Card>
    </div>
  )
}
