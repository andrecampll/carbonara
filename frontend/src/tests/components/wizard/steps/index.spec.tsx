import { MockedProvider } from '@apollo/client/testing'
import { cleanup, screen } from '@testing-library/react'
import { HTMLAttributes } from 'react'
import { afterEach, describe, expect, it, vitest } from 'vitest'

import { Steps } from '@/components/wizard/steps'
import { render } from '@/utils/tests/render'

vitest.mock('next/image', () => ({
  __esModule: true,
  default: (props: HTMLAttributes<HTMLImageElement>) => {
    return <img {...props} />
  },
}))

afterEach(cleanup)

describe('Steps', () => {
  it('should be able to render Steps component', () => {
    const { container } = render(
      <MockedProvider addTypename={false}>
        <Steps>
          <div>test</div>
        </Steps>
      </MockedProvider>,
    )

    expect(screen.getByText(/test/i)).toBeDefined()
    expect(container.firstChild).toMatchSnapshot()
  })
})
