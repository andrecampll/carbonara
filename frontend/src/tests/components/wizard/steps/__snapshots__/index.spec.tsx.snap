// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Steps > should be able to render Steps component 1`] = `
<div>
  test
</div>
`;
